Add allowlist for complex backward

anjali411 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45461 Add allowlist for complex backward**

This PR disables autograd for all C -> C, R -> C functions which are not included in the allowlist `GRADIENT_IMPLEMENTED_FOR_COMPLEX`. In practice, there will be a RuntimeError during forward computation when the outputs are differentiable:
```
>>> x=torch.randn(4, 4, requires_grad=True, dtype=torch.cdouble)
>>> x.pow(3)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: pow does not support automatic differentiation for outputs with complex dtype.
```

The implicit assumption here is that all the C -> R functions have correct backward definitions. So before merging this PR, the following functions must be tested and verified to have correct backward definitions:
`torch.abs` (updated in #39955 ), `torch.angle`, `torch.norm`, `torch.irfft`, `torch.istft`.

Differential Revision: [D23998156](https://our.internmc.facebook.com/intern/diff/D23998156)

anjali411 committed:
Add whitelist for complex backward

[ghstack-poisoned]

anjali411 commented on tools/autograd/gen_variable_type.py:
```diff
@@ -142,6 +142,19 @@
+    'eq_', 'ne_', 'add', '__radd__', 'sum', '_conj', 'sin', 'cos', 'mul', 'sinh',
+    'cosh', '__rmul__', 'sgn', 'view_as_real', 'real', 'imag', 'asin', 'acos', 'sub',
+    'div', 'cat', 'view_as_complex', 'neg', 'complex', 'select', '_s_where', 'as_strided',
+    '_fft_with_size'
```
cc. @peterbell10 @mruberry

peterbell10 replied:
`_fft_with_size` doesn't use complex, it uses (..., 2) shaped real tensors.

mruberry replied:
fft failures are real, though.

anjali411 committed:
Update on "Add whitelist for complex backward"


This PR disables autograd for all C -> C, R -> C functions which are not included in the whitelist `GRADIENT_IMPLEMENTED_FOR_COMPLEX`. In practice, there will be a RuntimeError during forward computation when the outputs are differentiable:
```
>>> x=torch.randn(4, 4, requires_grad=True, dtype=torch.cdouble)
>>> x.pow(3)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: pow does not support automatic differentiation for outputs with complex dtype.
```

The implicit assumption here is that all the C -> R functions have correct backward definitions. So before merging this PR, the following functions must be tested and verified to have correct backward definitions:
`torch.abs`, `torch.angle`, `torch.norm`, `torch.irfft`, `torch.istft`.

[ghstack-poisoned]

albanD commented on tools/autograd/gen_variable_type.py:
```diff
@@ -961,6 +974,16 @@ def setup_derivative(differentiable_inputs):
+        if base_name in GRADIENT_IMPLEMENTED_FOR_COMPLEX:
+            return body
+        for arg in differentiable_outputs:
+            if arg['type'] == 'Tensor':
```
What about TensorList?
In particular functions like `unbind()` will return such objects.

anjali411 replied:
The functions that are differentiable and return TensorList are: `torch.unbind`, `torch.split` (both of which have correct backward definition for complex). So, I think its ok to just not do anything for that case. However, I'll add check for TensorList type for any functions that maybe added in future.

albanD replied:
There are no other? Sounds good.
But yes a check is nice to make sure we don't break this in the future.

anjali411 replied:
There's `_cudnn_rnn_backward` but it's non-differentiable. 
Added `split', split_with_sizes, unsafe_split, split_with_sizes_backward` to the list.

albanD commented on torch/csrc/autograd/VariableTypeUtils.h:
```diff
@@ -67,6 +67,14 @@ inline void throw_error_out_requires_grad(const char* name) {
       "but one of the arguments requires grad.");
 }
 
+inline void throw_error_for_complex_fns_backward_not_implemented(const Tensor& tensor, const char* name) {
```
nit: name looks overly verbose.

albanD reviewed (commented):
I did not check the C->R functions nor the fact that the functions in the list are actually properly implemented.

The rest of the codegen looks good except the TensorList support.
Also it would be nice to show what the new generated code looks like for a sample function.

anjali411 committed:
Update on "Add whitelist for complex backward"


This PR disables autograd for all C -> C, R -> C functions which are not included in the whitelist `GRADIENT_IMPLEMENTED_FOR_COMPLEX`. In practice, there will be a RuntimeError during forward computation when the outputs are differentiable:
```
>>> x=torch.randn(4, 4, requires_grad=True, dtype=torch.cdouble)
>>> x.pow(3)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: pow does not support automatic differentiation for outputs with complex dtype.
```

The implicit assumption here is that all the C -> R functions have correct backward definitions. So before merging this PR, the following functions must be tested and verified to have correct backward definitions:
`torch.abs` (updated in #39955 ), `torch.angle`, `torch.norm`, `torch.irfft`, `torch.istft`.

[ghstack-poisoned]

mruberry commented on tools/autograd/gen_variable_type.py:
```diff
@@ -961,6 +974,18 @@ def setup_derivative(differentiable_inputs):
         body.append(SETUP_DERIVATIVE.substitute(env, setup=setup))
         return body
 
+    def emit_check_if_in_complex_autograd_whitelist():
```
"whitelist"->"allowlist"

mruberry replied:
(might as well change the title, too)

mruberry commented on torch/csrc/autograd/VariableTypeUtils.h:
```diff
@@ -67,6 +67,14 @@ inline void throw_error_out_requires_grad(const char* name) {
 }
 
+inline void throw_error_for_complex_fns_backward_not_implemented(const Tensor& tensor, const char* name) {
+  if (tensor.requires_grad() && tensor.is_complex()) {
```
Not a TORCH_CHECK here?

anjali411 committed:
Update on "Add whitelist for complex backward"


This PR disables autograd for all C -> C, R -> C functions which are not included in the whitelist `GRADIENT_IMPLEMENTED_FOR_COMPLEX`. In practice, there will be a RuntimeError during forward computation when the outputs are differentiable:
```
>>> x=torch.randn(4, 4, requires_grad=True, dtype=torch.cdouble)
>>> x.pow(3)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: pow does not support automatic differentiation for outputs with complex dtype.
```

The implicit assumption here is that all the C -> R functions have correct backward definitions. So before merging this PR, the following functions must be tested and verified to have correct backward definitions:
`torch.abs` (updated in #39955 ), `torch.angle`, `torch.norm`, `torch.irfft`, `torch.istft`.

[ghstack-poisoned]

anjali411 committed:
Update on "Add whitelist for complex backward"


This PR disables autograd for all C -> C, R -> C functions which are not included in the whitelist `GRADIENT_IMPLEMENTED_FOR_COMPLEX`. In practice, there will be a RuntimeError during forward computation when the outputs are differentiable:
```
>>> x=torch.randn(4, 4, requires_grad=True, dtype=torch.cdouble)
>>> x.pow(3)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: pow does not support automatic differentiation for outputs with complex dtype.
```

The implicit assumption here is that all the C -> R functions have correct backward definitions. So before merging this PR, the following functions must be tested and verified to have correct backward definitions:
`torch.abs` (updated in #39955 ), `torch.angle`, `torch.norm`, `torch.irfft`, `torch.istft`.

[ghstack-poisoned]

albanD reviewed (approved):
lgtm

ezyang commented:
Thanks, looks good.

cc @robieta I'm expecting a mild increase in instruction count here for AD benchmarks, will be more pronounced for operators on tensor lists.

mruberry commented:
I think the test failures are OK. I'm not sure why the XLA build is failing but it doesn't look related to this PR.

albanD commented:
@ezyang I think (almost) all the TensorList op we have are actually in the allowlist. So the impact should be minimal.

anjali411 committed:
Update on "Add whitelist for complex backward"


This PR disables autograd for all C -> C, R -> C functions which are not included in the whitelist `GRADIENT_IMPLEMENTED_FOR_COMPLEX`. In practice, there will be a RuntimeError during forward computation when the outputs are differentiable:
```
>>> x=torch.randn(4, 4, requires_grad=True, dtype=torch.cdouble)
>>> x.pow(3)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: pow does not support automatic differentiation for outputs with complex dtype.
```

The implicit assumption here is that all the C -> R functions have correct backward definitions. So before merging this PR, the following functions must be tested and verified to have correct backward definitions:
`torch.abs` (updated in #39955 ), `torch.angle`, `torch.norm`, `torch.irfft`, `torch.istft`.

Differential Revision: [D23998156](https://our.internmc.facebook.com/intern/diff/D23998156)

[ghstack-poisoned]

anjali411 committed:
Update on "Add whitelist for complex backward"


This PR disables autograd for all C -> C, R -> C functions which are not included in the whitelist `GRADIENT_IMPLEMENTED_FOR_COMPLEX`. In practice, there will be a RuntimeError during forward computation when the outputs are differentiable:
```
>>> x=torch.randn(4, 4, requires_grad=True, dtype=torch.cdouble)
>>> x.pow(3)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: pow does not support automatic differentiation for outputs with complex dtype.
```

The implicit assumption here is that all the C -> R functions have correct backward definitions. So before merging this PR, the following functions must be tested and verified to have correct backward definitions:
`torch.abs` (updated in #39955 ), `torch.angle`, `torch.norm`, `torch.irfft`, `torch.istft`.

Differential Revision: [D23998156](https://our.internmc.facebook.com/intern/diff/D23998156)

[ghstack-poisoned]

natalieklein commented:
Sorry to piggyback on a closed issue and I'm happy to open a new one if easier, but I figured those in this thread might be best able to help. I'm using Pytorch 1.10.0 and I'm having a problem using `torch.istft` with `return_complex=True` (I want a complex signal regenerated from the complex STFT). I am able to use `torch.istft` with the default `return_complex=False`, but with `return_complex=True` I get this exception: `RuntimeError: col2im does not support automatic differentiation for outputs with complex dtype.` I saw `col2im` is in the cpp code implementing `istft`, but I'm not sure how to address this error for the complex-output use case. Many thanks.
